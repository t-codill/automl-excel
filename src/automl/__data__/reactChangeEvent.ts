// tslint:disable: no-any
export const reactChangeEvent: React.ChangeEvent<any> = {
    target: {}
} as unknown as React.ChangeEvent<any>;
