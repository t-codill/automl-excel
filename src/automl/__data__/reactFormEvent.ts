// tslint:disable: no-any
export const reactFormEvent: React.FormEvent<any> = {
    preventDefault(): void { return; }
} as unknown as React.FormEvent<any>;
