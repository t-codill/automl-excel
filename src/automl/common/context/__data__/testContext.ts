import { PageNames } from "../../PageNames";
import { Logger } from "../../utils/logger";
import { IWorkspaceContextProps, WorkspaceError } from "../IWorkspaceProps";
import { WorkspaceFlight } from "../WorkspaceFlight";

export function getLogCustomEventSpy(): jest.SpyInstance<ReturnType<Logger["logCustomEvent"]>> {
    const spy = jest.spyOn(Logger.prototype, "logCustomEvent");
    spy.mockImplementation(jest.fn());
    return spy;
}
let pageName: PageNames = PageNames.Unknown;
export const testContext: IWorkspaceContextProps = {
    get pageName(): PageNames { return pageName; },
    get theme(): string { return "light"; },
    discoverUrls: {},
    flight: new WorkspaceFlight(""),
    location: "eastus",
    resourceGroupName: "testResource",
    subscriptionId: "00000000-0000-0000-0000-000000000000",
    logger: new Logger("test"),
    getToken: () => "testToken",
    setPageName: (value: PageNames) => { pageName = value; },
    workspaceName: "testWorkSpace",
    setNavigationBarButtons(): void { return; },
    onError: (err: WorkspaceError) => { throw err; },
    clearErrors: () => { return; },
    setHeader: () => { return; },
    setLoading: () => { return; }
};
