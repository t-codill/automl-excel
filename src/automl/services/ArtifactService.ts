import { ArtifactAPI, ArtifactAPIModels } from "@vienna/artifact";
import { blob2string } from "../common/utils/blob2string";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseNonArm } from "./ServiceBaseNonArm";

const artifactClientApiVersion = "1.0";
const runOrigin = "ExperimentRun";
const modelPath = "outputs/model.pkl";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export interface IRunDataForArtifact {
    runId: string | undefined;
    parentRunId: string | undefined;
    status: string | undefined;
}

async function getArtifactContent(artifact: ArtifactAPIModels.GetArtifactContentByIdResponse): Promise<string | undefined> {
    const blobBody = await artifact.blobBody;
    if (blobBody) {
        return blob2string(blobBody);
    }
    return undefined;
}
export class ArtifactService extends ServiceBaseNonArm<ArtifactAPI> {
    constructor(props: IServiceBaseProps) {
        super(props, ArtifactAPI, props.discoverUrls.history);
    }
    public async tryGetContentForRun(runId: string, path: string): Promise<string | null | undefined> {
        return this.tryGetContent({ origin: runOrigin, container: runId, path });
    }

    public async tryGetContent(artifact: IArtifact): Promise<string | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getContent(artifact: IArtifact): Promise<string | undefined> {
        return this.send(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return this.parallelGetAllValues(params, async (param, client, abortSignal) => {
            const artifact = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                param.origin,
                param.container,
                artifactClientApiVersion, {
                    path: param.path,
                    abortSignal
                });
            return getArtifactContent(artifact);
        });
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return this.parallelGetAllValuesWithContinuationToken(runs, async (runId, client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async getAllArtifacts(runId: string): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async tryGetArtifactUrl(artifact: IArtifact): Promise<ArtifactAPIModels.GetArtifactContentInformation2Response | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            return client.getArtifactContentInformation2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    abortSignal,
                    path: artifact.path
                });
        });
    }

    public async getModelUrl(run: IRunDataForArtifact): Promise<string | null | undefined> {
        if (!run || !run.runId || !run.parentRunId || run.status !== "Completed") {
            return null;
        }
        const result = await this.tryGetArtifactUrl({ container: run.runId, origin: runOrigin, path: modelPath });
        if (result === undefined) {
            return undefined;
        }
        if (!result || !result.contentUri) {
            return null;
        }
        return result.contentUri;
    }
}
