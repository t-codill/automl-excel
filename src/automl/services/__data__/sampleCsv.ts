export const sampleCsv = `a,b,c,d,e
1,2,3,4,5
1,2,3,4,5
1,2,3,4,5
1,2,3,4,5
1,2,3,4,5
`;

const largeCsvHeader = Array(50)
    .fill("aaaaa,bbbbb,ccccc,ddddd,eeeee")
    .join(",");

const largeCsvRow = Array(50)
    .fill("11111,22222,33333,44444,55555")
    .join(",");

export const largeCsv = `${largeCsvHeader}
${Array(50)
        .fill(largeCsvRow)
        .join("\n")}
`;
