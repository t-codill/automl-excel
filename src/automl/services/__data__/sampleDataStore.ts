import { AzureMachineLearningDatastoreManagementModels } from "@vienna/datastore";

export const sampleDataStore: Required<AzureMachineLearningDatastoreManagementModels.DataStoreDto> = {
    dataStoreType: "AzureBlob",
    name: "defaultDataStore",
    azureStorageSection: {
        accountKey: "key",
        accountName: "account",
        credentialType: "AccountKey",
        credential: "secret",
        containerName: "defaultContainer",
        isSas: false
    },
    azureDataLakeSection: {},
    azurePostgreSqlSection: {},
    azureSqlDatabaseSection: {},
    hasBeenValidated: true
};
