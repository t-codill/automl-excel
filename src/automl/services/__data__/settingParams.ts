import { AdvancedSetting } from "../AdvancedSetting";

export const settingParams: AdvancedSetting = {
    jobType: "classification",
    column: "1",
    metric: "AUC_weighted",
    experimentTimeoutMinutes: 60,
    maxIteration: 15,
    experimentExitScore: 0.99,
    maxConcurrent: 1,
    maxCores: 2,
    preprocess: true,
    nCrossValidations: 5,
    blacklistAlgos: null,
    validationSize: 0.2,
    grainColumns: null,
    maxHorizon: null,
    timeSeriesColumn: null
};
