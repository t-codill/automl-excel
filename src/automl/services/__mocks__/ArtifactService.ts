import { ArtifactAPIModels } from "@vienna/artifact";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export class ArtifactService {
    private readonly tryGetContentResult = "tryGetContent";
    private readonly getAllArtifactsResult: ArtifactAPIModels.ArtifactDto[] = [
        {
            artifactId: "artifactId1",
            container: "container1",
            origin: "origin1",
            path: "path1"
        },
        {
            artifactId: "artifactId2",
            container: "container2",
            origin: "origin2",
            path: "path2"
        }
    ];
    private readonly tryGetArtifactUrlResult: ArtifactAPIModels.ArtifactContentInformationDto = {
        contentUri: "contentUri",
        path: "path"
    };
    private readonly getModelUrlResult = "getModelUrl";
    private readonly getModelUrlsResult = ["getModelUrls1", "getModelUrls2"];

    public async tryGetContentForRun(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async tryGetContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return params.map((p) => `"getAllContentsResult:${p.origin}/${p.container}/${p.path}`);
    }

    public async getAllArtifacts(): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllArtifactsResult;
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return runs.map(() => this.getAllArtifactsResult);
    }

    public async tryGetArtifactUrl(): Promise<ArtifactAPIModels.ArtifactContentInformationDto | null | undefined> {
        return this.tryGetArtifactUrlResult;
    }

    public async getModelUrl(): Promise<string | undefined> {
        return this.getModelUrlResult;
    }

    public async getModelUrls(): Promise<Array<string | null> | undefined> {
        return this.getModelUrlsResult;
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
