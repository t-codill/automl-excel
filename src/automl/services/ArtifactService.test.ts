import { ArtifactAPI } from "@vienna/artifact";
import { restCanceledError } from "../../__data__/restCanceledError";
import { restNotFoundError } from "../../__data__/restNotFoundError";
import { testContext } from "../common/context/__data__/testContext";
import { ArtifactService } from "./ArtifactService";

jest.mock("@vienna/artifact");

const service: ArtifactService = new ArtifactService(testContext);

describe("ArtifactService", () => {
    it("should getAllArtifacts", async () => {
        const result = await service.getAllArtifacts("AutoML_00000000-0000-0000-0000-000000000000");
        expect(result)
            .toMatchSnapshot();
    });
    it("should getAllArtifactsForRuns", async () => {
        const result = await service.getAllArtifactsForRuns(["AutoML_00000000-0000-0000-0000-000000000000", "AutoML_00000000-0000-0000-0000-000000000001"]);
        expect(result)
            .toMatchSnapshot();
    });
    it("should getAllContents", async () => {
        const result = await service.getAllContents([{
            container: "sampleContainer",
            origin: "sampleOrigin",
            path: "samplePath"
        }]);
        expect(result)
            .toMatchSnapshot();
    });
    it("should tryGetArtifactUrl", async () => {
        const result = await service.tryGetArtifactUrl({
            container: "sampleContainer",
            origin: "sampleOrigin",
            path: "samplePath"
        });
        expect(result)
            .toMatchSnapshot();
    });
    it("should tryGetContentForRun", async () => {
        const result = await service.tryGetContentForRun("AutoML_00000000-0000-0000-0000-000000000000", "samplePath");
        expect(result)
            .toBe("sampleArtifactContent");
    });
    describe("getModelUrl", () => {
        it("should return null for non-completed runs", async () => {
            const result = await service.getModelUrl({
                parentRunId: "AutoML_00000000-0000-0000-0000-000000000000",
                runId: "AutoML_00000000-0000-0000-0000-000000000000_1",
                status: "Failed"
            });
            expect(result)
                .toBeNull();
        });

        it("should return value", async () => {
            const result = await service.getModelUrl({
                parentRunId: "AutoML_00000000-0000-0000-0000-000000000000",
                runId: "AutoML_00000000-0000-0000-0000-000000000000_1",
                status: "Completed"
            });
            expect(result)
                .toBe("contentUri");
        });

        it("should return null if model does not exist", async () => {
            const spy = jest.spyOn(ArtifactAPI.prototype, "getArtifactContentInformation2");
            spy.mockImplementation(() => { throw restNotFoundError; });
            const result = await service.getModelUrl({
                parentRunId: "AutoML_00000000-0000-0000-0000-000000000000",
                runId: "AutoML_00000000-0000-0000-0000-000000000000_1",
                status: "Completed"
            });
            expect(result)
                .toBeNull();
        });

        it("should return undefined if call is canceled", async () => {
            const spy = jest.spyOn(ArtifactAPI.prototype, "getArtifactContentInformation2");
            spy.mockImplementation(() => { throw restCanceledError; });
            const result = await service.getModelUrl({
                parentRunId: "AutoML_00000000-0000-0000-0000-000000000000",
                runId: "AutoML_00000000-0000-0000-0000-000000000000_1",
                status: "Completed"
            });
            expect(result)
                .toBeUndefined();
        });
        it("should get content", async () => {
            const result = await service.getContent({
                container: "sampleContainer",
                origin: "sampleOrigin",
                path: "samplePath"
            });
            expect(result)
                .toBe("sampleArtifactContent");
        });
    });
    describe("tryGetContent", () => {
        it("should get Value", async () => {
            const result = await service.tryGetContent({
                container: "sampleContainer",
                origin: "sampleOrigin",
                path: "samplePath"
            });
            expect(result)
                .toBe("sampleArtifactContent");
        });

        it("should be undefined if canceled", async () => {
            const spy = jest.spyOn(ArtifactAPI.prototype, "getArtifactContentById2");
            spy.mockImplementation(() => { throw restCanceledError; });
            const result = await service.tryGetContent({
                container: "sampleContainer",
                origin: "sampleOrigin",
                path: "samplePath"
            });
            expect(result)
                .toBeUndefined();
        });

        it("should be undefined without blobBody", async () => {
            const spy: jest.SpyInstance = jest.spyOn(ArtifactAPI.prototype, "getArtifactContentById2");
            spy.mockReturnValue(() => ({}));
            const result = await service.tryGetContent({
                container: "sampleContainer",
                origin: "sampleOrigin",
                path: "samplePath"
            });
            expect(result)
                .toBeUndefined();
        });
    });
});
