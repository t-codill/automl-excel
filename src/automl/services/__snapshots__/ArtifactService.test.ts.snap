// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtifactService should getAllArtifacts 1`] = `
Array [
  Object {
    "artifactId": "artifactId",
    "container": "container",
    "createdTime": 2019-01-01T00:00:00.000Z,
    "etag": "etag",
    "origin": "origin",
    "path": "path",
  },
]
`;

exports[`ArtifactService should getAllArtifactsForRuns 1`] = `
Array [
  Array [
    Object {
      "artifactId": "artifactId",
      "container": "container",
      "createdTime": 2019-01-01T00:00:00.000Z,
      "etag": "etag",
      "origin": "origin",
      "path": "path",
    },
  ],
  Array [
    Object {
      "artifactId": "artifactId",
      "container": "container",
      "createdTime": 2019-01-01T00:00:00.000Z,
      "etag": "etag",
      "origin": "origin",
      "path": "path",
    },
  ],
]
`;

exports[`ArtifactService should getAllContents 1`] = `
Array [
  "sampleArtifactContent",
]
`;

exports[`ArtifactService should tryGetArtifactUrl 1`] = `
Object {
  "_mockParameters": Array [
    "00000000-0000-0000-0000-000000000000",
    "testResource",
    "testWorkSpace",
    "sampleOrigin",
    "sampleContainer",
    "1.0",
    Object {
      "abortSignal": Object {},
      "path": "samplePath",
    },
  ],
  "container": "container",
  "contentUri": "contentUri",
  "origin": "origin",
  "path": "path",
}
`;
